### ISSUE TITLE
Illegal instruction, not sure how it occurs, rr report included

### ISSUE THREAD

**Issue by bhalonen** [JuliaLang/julia#38312]
https://s3.amazonaws.com/julialang-dumps/reports/2020-11-05T14-56-56-bhalonen.tar.zst

This is for last nights build, though it happens on 1.5.2 and 1.5.1 as well.
```
signal (4): Illegal instruction
in expression starting at /home/brenthalonen/projects/coho-deploy/CohoSim.jl/test/acmt.jl:1
getproperty at ./Base.jl:33 [inlined]
rehash! at ./dict.jl:178
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2191 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2373
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1693 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:115
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:204
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:434
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:489
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:489
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:669
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:879
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:827
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:931
eval at ./boot.jl:360 [inlined]
include_string at ./loading.jl:1082
```

```
julia> versioninfo()
Julia Version 1.6.0-DEV.1426
Commit b00802bc5f (2020-11-04 22:35 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.0 (ORCJIT, skylake)
```

**Comment by bhalonen** [comment]
Put a few hours into extracting a mwe but gave up.

**Comment by bhalonen** [comment]
There is a struct with a union over missing involved might be similar to #38224

**Comment by bhalonen** [comment]
the bug does not occur in 1.3, additional info

**Comment by Keno** [comment]
Looks like the issue is a concrete type that doesn't get interned:

```
(rr) p jl_(0x7fb3a86e9cf0)
Array{Intervals.Interval{var"#s13", L, R} where R<:Intervals.Bound where L<:Intervals.Bound where var"#s13"<:(Union{Unitful.Quantity{T, Unitful.Dimensions{(Unitful.Dimension{:Time}(power=Base.Rational{Int64}(num=1, den=1)),)}(), U}, Unitful.Level{L, S, Unitful.Quantity{T, Unitful.Dimensions{(Unitful.Dimension{:Time}(power=Base.Rational{Int64}(num=1, den=1)),)}(), U}} where S where L} where U where T), 1}
$58 = void
(rr) p jl_(0x7fb3a33fc550)
Array{Intervals.Interval{var"#s13", L, R} where R<:Intervals.Bound where L<:Intervals.Bound where var"#s13"<:(Union{Unitful.Quantity{T, Unitful.Dimensions{(Unitful.Dimension{:Time}(power=Base.Rational{Int64}(num=1, den=1)),)}(), U}, Unitful.Level{L, S, Unitful.Quantity{T, Unitful.Dimensions{(Unitful.Dimension{:Time}(power=Base.Rational{Int64}(num=1, den=1)),)}(), U}} where S where L} where U where T), 1}
```

Codegen gets very unhappy as a result.

**Comment by Keno** [comment]
Alright, both these types were deserialized from (different) .ji files. @JeffBezanson @vtjnash how is uniqueing of concrete types supposed to work for datatypes deserialized from .ji files.

**Comment by Keno** [comment]
MWE:
```
# Issue #38312
let
    TheType = """Array{Ref{Val{1}}, 1}"""
    load_path = mktempdir()
    load_cache_path = mktempdir()
    try
        write(joinpath(load_path, "Foo38312.jl"),
            """
            module Foo38312

            const TheType = $TheType

            end
            """)
        write(joinpath(load_path, "Bar38312.jl"),
            """
            module Bar38312

            const TheType = $TheType

            end
            """)

        pushfirst!(LOAD_PATH, load_path)
        pushfirst!(DEPOT_PATH, load_cache_path)

        Base.compilecache(Base.PkgId("Foo38312"))
        Base.compilecache(Base.PkgId("Bar38312"))
        @test pointer_from_objref((@eval (using Foo38312; Foo38312)).TheType) ===
        pointer_from_objref((@eval (using Bar38312; Bar38312)).TheType)
    finally
        rm(load_path, recursive=true)
        rm(load_cache_path, recursive=true)
    end
end
```

**Comment by Keno** [comment]
Huh:
```
(rr) p jl_(tj)
Ref{Base.Val{1}}
$104 = void
(rr) p jl_(kj)
Ref{Base.Val{1}}
$105 = void
(rr) p jl_types_equal(tj, kj)
$106 = 0
```

**Comment by Keno** [comment]
Proposed fix in #38339 .
